### PR TITLE
Chore: update dependencies

### PR DESCRIPTION
**Pull request by kevinwcyu**
**What this PR does / why we need it**:
Updating dependencies

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/oss-plugin-partnerships/issues/922
Fixes https://github.com/grafana/oss-plugin-partnerships/issues/923

**Special notes for your reviewer**:

I've organized the changes into different commits to make it easier to review:
- https://github.com/grafana/opensearch-datasource/pull/476/commits/daeb8ca5a115f7eecd3df33c474a24cd0e4bf89e:
  - mainly automated changes from running `run npx @grafana/create-plugin@latest update`. The `.config/` directory changes are all automated
  - manual changes include updating node to version 20 in `.nvmrc`, removing `TextEncoder` `jest-setup.js` as it's now included in `.config/jest-setup.js`, and updating dependencies in `package.json`
- https://github.com/grafana/opensearch-datasource/pull/476/commits/3e19ca823479d5cd4c475e4272cdc4930817bf8f:
  - updateg go dependencies to their latest version and ran `go mod tidy` which auto updated `go` and `toolchain` in `go.mod`
- https://github.com/grafana/opensearch-datasource/pull/476/commits/968443b059f28f494e70942e72866250d64bf09a:
  - `ArrayVector` is deprecated and causes an error when it's used, the fix is to replace usages of it with plain arrays
- https://github.com/grafana/opensearch-datasource/pull/476/commits/9885593564118c729d37f50d6e968de9e199b053:
  - `renderHooks` now comes from `@testing-library/react` instead of `@testing-library/react-hooks`
  - `import '@testing-library/jest-dom';` to fix a typescript error where `toBeInTheDocument` is not found
  - `src/components/QueryEditor/MetricAggregationsEditor/SettingsEditor/index.test.tsx` I believe there were some changes in the `Select` component from `@grafana/ui` v10.x.x to v11.x.x so this just updated the test to account for the change in the structure of the `Select` options
- https://github.com/grafana/opensearch-datasource/pull/476/commits/ed68606f4136f223e3e2d89332a9dec823ebe377:
  - replace our own `Action` type in favour of using the one from `@reduxjs/toolkit`
- See the note below on `@reduxjs/toolkit updates` for a description of the following changes
  - https://github.com/grafana/opensearch-datasource/pull/476/commits/d53d84917358b8915b254bdb5c392c6be9f59318
  - https://github.com/grafana/opensearch-datasource/pull/476/commits/52c27f35da89b2ac690b8b91d2cfc88e5928fadb
  - https://github.com/grafana/opensearch-datasource/pull/476/commits/7a260158b88373ac868a5e11b41a8a840f2559d0
  - https://github.com/grafana/opensearch-datasource/pull/476/commits/144621fc0e52ac66e4843b2e78dd4c1ade3e6eaf
  - https://github.com/grafana/opensearch-datasource/pull/476/commits/081f70e965a349430d2d6c9d6fe0ea8fad1e818f
- https://github.com/grafana/opensearch-datasource/pull/476/commits/a7201325fdad409a83d203c9f2aaacd5aea1944c:
  - updates the dependabot config to open PRs on a weekly basis and group PRs by package-ecosystem.
- https://github.com/grafana/opensearch-datasource/pull/476/commits/132080d36e8c6285ba59eda07ffdd9bab03b50a3:
  - bump frontend dependencies after merging main
- https://github.com/grafana/opensearch-datasource/pull/476/commits/5b3766184a04359e46f9c4b4945f0f00e5d9e059
  - bump backend dependencies after merging main

**@reduxjs/toolkit updates**
After updating `@reduxjs/toolkit` there were a lot of type errors with the actions/reducers. Before the update, we manually defined the action types and action creators. Something about the manually created action types was clashing with the updated types so I ended up removing the manually defined action types and action creators and used the `createAction` help from `@reduxjs/toolkit` which is a helper for creating an action creator that can also be used in place of the action type. This allowed me to delete the manually created action types and action creators.

Here's an example of the changes to the `changeBucketAggregationType` action to make it a bit more concrete

**Previously**
```ts
/* src/components/QueryEditor/BucketAggregationsEditor/state/types.ts */
// action creator is manually created and we need to specify the payload it returns and make sure to set the type to `CHANGE_BUCKET_AGG_TYPE`
export const changeBucketAggregationType = (
  id: BucketAggregation['id'],
  newType: BucketAggregation['type']
): BucketAggregationAction => ({
  type: CHANGE_BUCKET_AGG_TYPE,
  payload: {
    id,
    newType,
  },
});

/* src/components/QueryEditor/BucketAggregationsEditor/state/types.ts */
// action type `ChangeBucketAggregationTypeAction` defined
export interface ChangeBucketAggregationTypeAction extends Action<typeof CHANGE_BUCKET_AGG_TYPE> {
  payload: {
    id: BucketAggregation['id'];
    newType: BucketAggregation['type'];
  };
}

// union type for all bucket aggregation action types
export type BucketAggregationAction<T extends BucketAggregation = BucketAggregation> =
  | AddBucketAggregationAction
  | RemoveBucketAggregationAction
  | ChangeBucketAggregationTypeAction // `ChangeBucketAggregationTypeAction` needs to be included here
  | ChangeBucketAggregationFieldAction
  | ChangeBucketAggregationSettingAction<T>;

/* src/components/QueryEditor/BucketAggregationsEditor/state/reducer.ts */
// reducer creator needs to include the `BucketAggregationAction` union type in its function signature
  (
    state: OpenSearchQuery['bucketAggs'],
    action: BucketAggregationAction | ChangeMetricTypeAction | InitAction
  ): OpenSearchQuery['bucketAggs'] => {
// the reducer needs to check the `action.type` that the action creator includes in its payload to make sure it matches `CHANGE_BUCKET_AGG_TYPE` that is set in the action creator
switch (action.type) {
  // ... other code excluded
  case CHANGE_BUCKET_AGG_TYPE:
```

**After updates**
```ts
/* src/components/QueryEditor/BucketAggregationsEditor/state/actions.ts */
// `createAction just needs `CHANGE_BUCKET_AGG_TYPE` to identify the action and we just pass the type for the expected payload
export const changeBucketAggregationType = createAction<{
  id: BucketAggregation['id'];
  newType: BucketAggregation['type'];
}>(CHANGE_BUCKET_AGG_TYPE);

/* src/components/QueryEditor/BucketAggregationsEditor/state/reducer.ts */
// the reducer creator can just use the `Action` type from `@reduxjs/toolkit` in its function signature instead of us needing to create a separate union type of all the bucket aggregation action types
// the reducer can just import the action creator we created above and use the `match` function to check if the action is the type we expect. This acts as a type predicate so we also get properly typed payloads in the reducer.
if (changeBucketAggregationType.match(action)) {
```

This is the bulk of the changes to the actions/reducers. The logic in the reducers is the exact same as before.

The only other thing to note about the action creators is that they now take a single argument for its payload. E.g.

**Previously**
```ts
export const changeBucketAggregationType = (
  id: BucketAggregation['id'],
  newType: BucketAggregation['type']
): BucketAggregationAction => ({
  type: CHANGE_BUCKET_AGG_TYPE,
  payload: {
    id,
    newType,
  },
});

// example usage

changeBucketAggregationType('newId', 'newType'); // <- multiple arguments
```

**After updates**
```ts
export const changeBucketAggregationType = createAction<{
  id: BucketAggregation['id'];
  newType: BucketAggregation['type'];
}>(CHANGE_BUCKET_AGG_TYPE);

// example usage
changeBucketAggregationType( { id: 'newId', newType: 'some-other-type' }) // <- pass arguments as an object
```

### CODE REVIEW

**Review by iwysiu — APPROVED** [review]
Nice!

File: src/hooks/useStatelessReducer.ts [hunk 1]
```diff
@@ -1,9 +1,7 @@
+import { Action } from '@reduxjs/toolkit';
+export { Action } from '@reduxjs/toolkit';
```
Comment (idastambuk) [inline src/hooks/useStatelessReducer.ts:2]:
```suggestion
```

**Review by idastambuk — APPROVED** [review]
Amazing work!

File: src/hooks/useStatelessReducer.ts (hunk 1 above)
Comment (kevinwcyu) [reply src/hooks/useStatelessReducer.ts]: 👍 I removed exporting it here, and just updated the rest of the files to import directly from `@reduxjs/toolkit`

File: src/components/QueryEditor/MetricAggregationsEditor/SettingsEditor/index.test.tsx [hunk 2]
```diff
@@ -76,10 +76,8 @@ describe('Settings Editor', () => {
       expect(settingsButtonEl).toBeInTheDocument();
       expect(settingsButtonEl.textContent).toBe(`Size: ${newSizeValue}, Order: ${initialOrder}`);
 
-      let select = (await screen.findByText('Order')).nextSibling!.firstChild!;
+      let select = await screen.findByLabelText('Order');
       await fireEvent.keyDown(select, { keyCode: 40 });
-      const scs = screen.getAllByLabelText('Select option');
-      expect(scs).toHaveLength(2);
```
[the remark this answers is not in the document]
Comment (kevinwcyu) [reply src/components/QueryEditor/MetricAggregationsEditor/SettingsEditor/index.test.tsx:82]: Updated!

**Pull request merged**